tor-android-binary/src/main/java/org/torproject/jni/TorService.java

```diff
@@ -346,6 +359,10 @@ public void run() {
                                 "--LogMessageDomains", "1",
                                 "--TruncateLogFile", "1"
                         ));
+                        if (owningProcessPid != -1) {
+                            lines.add("--__OwningControllerProcess");
+                            lines.add(Integer.toString(owningProcessPid));
+                        }
```

Comment:
This is to ensure that Tor doesn't outlive the app (for more than 15sec) if it crashes. Otherwise, it would continue idling on the system (potentially indefinitely) consuming resources.

Reply:
Android has native mechanisms for ensuring that a `Service` does not outlive its app.  Binding is one of them.  `TorService` works much more in the Android space than tor daemon ever did.  This kind of thing seems like a holdover from running tor as a daemon.  (If we eliminate the control port socket, then it will be even more in that direction.)

Reply:
I [read up](https://torproject.gitlab.io/torspec/control-spec.html#takeownership) a bit on this option, it doesn't seem like something that should be built into `TorService`.  If it is something you want to use, you can generate your own _torrc_ and `TorService` will use that file.

Reply:
We use this option in Briar to prevent the Tor process from outliving the controller process, which was a recurrent source of bugs for both Briar and Orbot in the past. When `owingProcessPid != -1` the controller and Tor library are running in different processes, so that possibility resurfaces.

As far as I know, Android doesn't guarantee that a service will be destroyed if a bound client in another process crashes. The service may be cached, for example. So I don't think it's wise to rely on Android destroying the service.

Reply:
Another quick thought here: when TorService is running as a foreground service (which we're doing in OnionShare because Android gives various power management advantages to foreground services), we've found than simply unbinding is not enough to stop or destroy the service. I don't know whether the same applies when a bound client dies, rather than unbinding, but I think it would at least be worth checking that the service gets destroyed before we stop using `__OwningControllerProcess`.

Reply:
> I had hoped to have the Binder stuff nicely documented with example code by now, but unfortunately it hasn't happened yet. But I can help with any issues that arise. #64
 
Thanks, but #64 won't address our use case. For OnionShare we need the service to keep running and enjoying the power management benefits of a foreground service even when the associated activity is in the background. A foreground service is the appropriate tool for this use case.

> If you're setting a Service to be a foreground service with a notification, then you have told Android that you are managing the lifecycle entirely. So the Binder reference counting shutdown won't take effect since it has been overridden.

Ah, that makes sense and matches what we've observed, thanks.

So given that the service won't stop when the client unbinds, we need a way of cleaning up the service in case the client process crashes or is killed. `__OwningControllerProcess` was designed to ensure that the Tor process exits if the controller process dies, which is the situation we're looking at here, so this seems like an appropriate place to use it.

It's true that we could add this option to `torrc` ourselves, but that would mean generating `torrc` dynamically just to insert this option which could more easily be added to the command line.

Reply:
> Thanks, but #64 won't address our use case. For OnionShare we need the service to keep running and enjoying the power management benefits of a foreground service even when the associated activity is in the background. A foreground service is the appropriate tool for this use case.

@grote said you want separate processes for the app and `TorService`.  So the Android-style architecture here would be to stick the notification on the task that needs to kept alive.  In OnionShare, that's the upload service.  Then the upload service binds to `TorService`, so `TorService` also gets foreground priority without having to put up a second notification.

Reply:
> So the Android-style architecture here would be to stick the notification on the task that needs to kept alive. In OnionShare, that's the upload service. Then the upload service binds to TorService, so TorService also gets foreground priority without having to put up a second notification.

This would add unnecessary complexity for us and any other app that wants to use this library to provide a hidden service, but yes, it seems like it would function.

Reply:
I cannot see why.  Service binding is not complex or hard to use.  And it is well documented.  And it is a documented way to tell Android about the app's lifecycle requirements.

Reply:
>  Then the upload service binds to `TorService`, so `TorService` also gets foreground priority without having to put up a second notification.

So I tried this. The system is restarting the upload service after the app crashed resulting in Tor running without the app. This is one of the issues this PR is meant to address and was tested to address properly.

```
W/ActivityManager: Scheduling restart of crashed service org.onionshare.android/.tor.UploadService in 1000ms for start-requested
```

Reply:
Where can I see this code?

Reply:
> It adds complexity because now we need two services: TorService itself, and a second service whose only purpose is to bind to TorService in order to ensure that TorService gets destroyed if the client dies. With __OwningControllerProcess TorService would be responsible for making sure it gets destroyed if the client dies, which seems like a more convenient library to me.

I agree that it does not make sense to use an additional service in the scenario you describe.  But that's not the scenario of neither OnionShare nor Briar as far as I understand them.  Both of those have other tasks that rely on `TorService`.  Those other tasks should be in a `Service` instance that is separate from `TorService`.  OnionShare could have multiple uploads running at the same time.  Each of those would run in an instance of `UploadService` that is bound to OnionShare's unique `TorService`.  Or there could be a `CoordinatorService` which manages the notifications and is set as a foreground service.  `CoordinatorService` binds to `TorService` and each `UploadService` so they all get foreground status.  `Service` subclasses are lightweight because they are designed with this kind of structure in mind.  

The goal here is to move tor-based services to work well with [`WorkManager`](https://developer.android.com/topic/libraries/architecture/workmanager/basics).  That ties into the Android power/network management and scheduling.  Here's their nutshell description:
* WorkManager is an Android library that runs deferrable background work when the work’s constraints are satisfied.
* WorkManager is intended for tasks that require a guarantee that the system will run them even if the app exits.

Reply:
I don't think we'll want to use WorkManager for OnionShare, as the tasks started by the app are expected to start immediately, not to be scheduled for later execution when some constraints are met.

> Those other tasks should be in a Service instance that is separate from TorService. OnionShare could have multiple uploads running at the same time. Each of those would run in an instance of UploadService that is bound to OnionShare's unique TorService. Or there could be a CoordinatorService which manages the notifications and is set as a foreground service. 

That's one possible way of structuring things, yes. But it would be nice if the library didn't impose a specific architecture on the application.

Reply:
WorkManager now considers foreground processing a first class use case: https://www.droidcon.com/2021/11/17/workmanager-back-to-the-foreground/ see at 6:30.  WorkManager is clearly the way forward with Tor on Android.

>That's one possible way of structuring things, yes. But it would be nice if the library didn't impose a specific architecture on the application.

This library does not impose that kind of architecture, Android does.  A central goal of this library is to make Tor behave natively on Android. Android works quite differently than UNIX, Windows, and macOS.  Using desktop lifecycle constructs on Android and iOS works quite poorly, that's why this is a goal.

Reply:
If you refer to `Expedited work` then that is "[for tasks that are important to the user and which complete within a few minutes.](https://developer.android.com/topic/libraries/architecture/workmanager#expedited)" This not the case for the OnionShare use-case where you might want to leave the share running for more than a day without the fear of yet another system service killing you early.

Reply:
From the link you sent:
> While your app is in the foreground, quotas won’t limit the execution of expedited work.

Here's more on using WorkManager for long running services, "Example use-cases for this new feature include bulk uploads or downloads":
https://developer.android.com/topic/libraries/architecture/workmanager/advanced/long-running

(_github messed up the threading so this is a repost_)